api/debug.go

```diff
@@ -95,3 +130,241 @@ func (d *DebugAPI) TraceBlockByHash(
 
 	return results, nil
 }
+
+func (d *DebugAPI) TraceCall(
+	ctx context.Context,
+	args TransactionArgs,
+	blockNrOrHash rpc.BlockNumberOrHash,
+	config *tracers.TraceCallConfig,
+) (interface{}, error) {
+	if err := rateLimit(ctx, d.ratelimiter, d.logger); err != nil {
+		return nil, err
+	}
+
+	txEncoded, err := encodeTxFromArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Default address in case user does not provide one
+	from := gethCommon.Address{}
+	if args.From != nil {
+		from = *args.From
+	}
+
+	var traceConfig *tracers.TraceConfig
+	if config != nil {
+		traceConfig = &config.TraceConfig
+	}
+
+	tracer, err := tracerForReceipt(traceConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	height, err := d.resolveBlockNumberOrHash(&blockNrOrHash)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := d.blocks.GetByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+
+	blockExecutor, err := d.executorAtBlock(block)
+	if err != nil {
+		return nil, err
+	}
+
+	tx := &gethTypes.Transaction{}
+	if err := tx.UnmarshalBinary(txEncoded); err != nil {
+		return nil, err
+	}
+
+	// call tracer
+	head := &gethTypes.Header{
+		Number: big.NewInt(int64(block.Height)),
+		Time:   block.Timestamp,
+	}
+	emulatorConfig := emulator.NewConfig(
+		emulator.WithChainID(d.config.EVMNetworkID),
+		emulator.WithBlockNumber(head.Number),
+		emulator.WithBlockTime(head.Time),
+	)
+	random := block.PrevRandao
+	stateDB := blockExecutor.StateDB
+	tracer.OnTxStart(
+		&tracing.VMContext{
+			Coinbase:    d.config.Coinbase,
+			BlockNumber: head.Number,
+			Time:        head.Time,
+			Random:      &random,
+			GasPrice:    d.config.GasPrice,
+			ChainConfig: emulatorConfig.ChainConfig,
+			StateDB:     stateDB,
+		},
+		tx,
+		from,
+	)
+
+	err = applyStateOverrides(config, stateDB)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := blockExecutor.Call(from, txEncoded, tracer)
+	if err != nil {
+		return nil, err
+	}
+
+	// call tracer on tx end
+	if tracer.OnTxEnd != nil {
+		tracer.OnTxEnd(result.Receipt(), result.ValidationError)
```

Comment:
I don't have a definitive answer, to be honest. But that's what's used in flow-go as well, regarding traces: https://github.com/onflow/flow-go/blob/master/fvm/evm/emulator/emulator.go#L191. This part has been moved to flow-go as well, so it will be removed once we get a version with that change included.